File: solutions/t9-spelling/go/phoneTranslate.go
package main

import (
	"fmt"
	"bufio"
	"os"
	"strings"
	"strconv"
)

func check(e error) {
    if e != nil {
        panic(e)
    }
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	var path string
    fmt.Println("Please insert the file name")
	for scanner.Scan() {
		path = scanner.Text();
		break;
	}

	file, err := os.Open(path)
    check(err)
  	defer file.Close()
	
	var content []string
    scanner = bufio.NewScanner(file)
	for scanner.Scan() {
	  	content = append(content, scanner.Text())
	}
    phoneTranslate(content)
}

func phoneTranslate(content []string){
	mapping := []string{" ", "" , "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}
	N,_:=strconv.ParseInt(content[0],0,64)
	var lastChar int
	for i:= int64(0); i<N; i++{
		word:= content[i+1]
		chars:= strings.Split(word,"")

		fmt.Print("Case #", i+1, ": ")
		for j:= 0; j<len(chars); j++{
			for k:= 0; k<len(mapping); k++{
				if(strings.Contains(mapping[k],chars[j])){
					if lastChar==k{
						fmt.Print(" ")	
					}
					for l:= 0; l<=strings.Index(mapping[k],chars[j]); l++{
						fmt.Print(k)
					}
					lastChar=k
				}
			}
		}	
		fmt.Print("\n")
	}
}

File: solutions/t9-spelling/dart/phoneTranslate.dart
import 'dart:io';
import 'dart:collection';
void main() async{
	print("Please input the file name on the same folder");
	var fileName = stdin.readLineSync();
	var config = new File(fileName);
	config.readAsLines().then(translate);
}

translate(List<String> lines){
	var N=0;
	var value = 0, index = 0, tempIndex = 0;
	var word="";
	var currentCase = 1;
	var repeated = false;
	var mapping = [" " , "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"];
	for (var line in lines){
		word = "";
		if(N==0)
			N = int.parse(line);
		else{
			for(var char in line.split('')){
				var element = mapping.firstWhere((o) => o.contains(char), orElse: () => '');
				tempIndex = mapping.indexOf(element);
				if(tempIndex == index)
					repeated = true;
				index = tempIndex;
				value = element.indexOf(char);
				for (var i = 0; i<=value; i++){
					if(repeated){
						word += " ";
						repeated = false;
					}
					word+="$index";
				}
			}
			print ("Case #$currentCase: $word");
			currentCase++;
		}
	}
}